Added external link and title to search promotions

TopDevPros opened the pull request:
Fixes #10725

Based on an earlier PR https://github.com/wagtail/wagtail/pull/4560

Users can now create a search pick for a Page **or** an external link. 

Added tests, doc strings, and updated docs to show how to use the new fields in a template.

Also, enabled the Query.get_most_popular() function to recognize the date_since parameter.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -1,3 +1,10 @@
+## To do:
+    - Eliminate duplicates of Query and QueryDailyHits classes in wagtail.search
+"""
+
```
No need to add this docstring at the top, we don't really do this. 

If there is something else to improve, best to just raise a new issue. Todo comments tend to get lost in the noise.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -37,8 +60,11 @@ def hits(self):
     def garbage_collect(cls):
         """
-        Deletes all Query records that have no daily hits or editors picks
+        # Deletes all Query records that have no daily hits or editors picks
```
The `#` in front of lines is not really necessary, we don't really do that in any other code.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -50,22 +76,54 @@ def garbage_collect(cls):
+            - First Query object matching query_string.
+
+        ## To do:
+            - Rename get() to get_or_create()
```
Same here, please remove the Todo. If you feel it's a critical issue we can add two methods, one with the new name and the current name as an alias. 

Then deprecate the old name in a future release.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -50,22 +76,54 @@ def garbage_collect(cls):
+
+        if date_since:
+            objects = objects.filter(daily_hits__date__gte=date_since)
+
```
Thank you for this, however, it would be a bit easier to review if this was in a standalone PR. Would you be ok to pull this change out to a different PR?

lb- replied:
If possible, can you pull out these changes to a different PR? I think this change is best reviewed on its own.

I see you are using `main` for this PR so you may need to create a new branch off main locally and call it 'backup' or similar. That branch will contain all the work on `main`.

Then, switch back to the `main` branch and remove these changes.

Alternatively, if you want to keep pushing for this change in the same PR, it will need discrete unit tests added in wagtail/search/tests/test_queries.py for this work, along with maybe some data points to test against.

It's probably much simpler to stay focused on one feature per PR though, even though I know it's nice to try to fix some "TODO" while in the code.

**Update:** Looks like there are tests, thank you, maybe they can move to the file noted above that appears to be specifically for testing the queries.

lb- commented on wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotion_form.html:
```diff
@@ -22,7 +22,23 @@
+
+        <strong>
+            {% trans "Or an external link" %}
+        </strong>
```
We may need to revisit this content, breaking up the translation parts like this will be difficult to work with for some languages. 

Plus, this content is likely going to need some refinement for accessibility. 

I'll run this locally and try to give some suggestions. 

For now - we should at least remove the empty paragraph with the space in it. CSS should be used if we need to do layout things.

lb- replied:
OK here is a recommended set of HTML for this area, the wording is a bit verbose but I think it may be a better place to start. Essentially using simpler semantic HTML basics (fieldset/legend) to break up the form. This also will work much better for screen readers.

I will reach out to the rest of the core team for any additional guidance on wording here but hopefully this will be a step in the right direction.

```html
    {% panel id=panel_id heading=heading heading_size="label" header_controls=header_controls %}
        {% if form.non_field_errors %}
            <ul>
                {% for error in form.non_field_errors %}
                    <li class="error-message">
                        <span>{{ error|escape }}</span>
                    </li>
                {% endfor %}
            </ul>
        {% endif %}
        <fieldset>
            <legend>{% trans "Use an internal Page for this Promotion" %}</legend>
            {% include "wagtailadmin/shared/field.html" with field=form.page only %}
        </fieldset>
        <fieldset>
            <legend>{% trans "Alternatively, use an external link for this Promotion" %}</legend>
            {% include "wagtailadmin/shared/field.html" with field=form.external_link only %}
            {% include "wagtailadmin/shared/field.html" with field=form.external_title only %}
        </fieldset>
        {% include "wagtailadmin/shared/field.html" with field=form.description only %}
    {% endpanel %}
```

<img width="1958" alt="Screenshot 2023-08-24 at 7 56 35 am" src="https://github.com/wagtail/wagtail/assets/1396140/acb0eb1f-ff10-44ba-99a4-dd9bc3605ead">

lb- commented on wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotions_formset.html:
```diff
@@ -11,7 +11,7 @@
     <div class="w-mb-4 -w-ml-4">
         <button type="button" class="button button-small button-secondary chooser__choose-button" id="id_{{ formset.prefix }}-ADD" value="Add">
-            {% icon name="plus-inverse" %}{% trans "Add recommended page" %}
+            {% icon name="plus-inverse" %}{% trans "Add a search result" %} {# not just a wagtail 'page' #}
```
Maybe... 'Add a recommend result'?

lb- reviewed (changes requested):
A solid start, thank you!

I have some feedback, some is a bit nitpicky but just trying to help this code align with other conventions. 

Hopefully this helps move your PR along. 

Please can you also review the CI failures, not sure if they're related to your changes. 

I'll run locally and give some feedback on the form layout when I get the chance. 

Again, really appreciate you taking the time to contribute.

lb- commented:
Will aim to another review, thanks for working through the comments.

lb- commented:
Still seeing some failures on the CI.

```
FAIL: test_post_reorder (wagtail.contrib.search_promotions.tests.TestSearchPromotionsEditView)
Change the order of the search promotions and verify the change.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.12/x64/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.10.12/x64/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/opt/hostedtoolcache/Python/3.10.12/x64/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/runner/work/wagtail/wagtail/wagtail/contrib/search_promotions/tests.py", line 590, in test_post_reorder
    self.assertEqual(Query.get("Hello").editors_picks.all()[0], self.search_pick)
  File "/opt/hostedtoolcache/Python/3.10.12/x64/lib/python3.10/unittest/case.py", line 845, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/hostedtoolcache/Python/3.10.12/x64/lib/python3.10/unittest/case.py", line 838, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: SearchPromotion(query="hello", page="Welcome to your new Wagtail site!") != SearchPromotion(query="hello", page="Root")

----------------------------------------------------------------------
Ran 6019 tests in 623.731s
```

Only on some Django version combinations though @TopDevPros

TopDevPros commented:
Thanks for the quick review, @LB.

CI bug?

We don't see

    SearchPromotion(query="hello", page="Welcome to your new Wagtail site!") != SearchPromotion(query="hello", page="Root")

anywhere in the code we have. It's definitely not in our search_promotions/test.py. The traceback line numbers don't match up.

It looks like the code tested by CI is not what we pushed. If you can point us to it, we're happy to take a look.

zerolab commented:
@TopDevPros https://github.com/wagtail/wagtail/blob/d616d481157e64919d395741a922afae8221eada/wagtail/contrib/search_promotions/tests.py#L585-L591 is the failing bit

lb- commented on wagtail/contrib/search_promotions/forms.py:
```diff
@@ -45,12 +61,29 @@ class Meta:
+        - minimum_forms (int): Number of forms required for this set.
+    """
+
+    REQUIRED_FIELDS_MISSING = _("You must recommend a related page OR an external url.")
```
Please update all user facing messages to use `URL` not `url`, this is how we refer to URLs in all other messaging currently.

However, it's probably better to consistently refer to these as 'external links', and then the inner parts are either `URL` or `Title`. Please update to something like this...

```
_("You must recommend a related page OR an external link.")
``

lb- commented on wagtail/contrib/search_promotions/forms.py:
```diff
@@ -45,12 +61,29 @@ class Meta:
+
+    REQUIRED_FIELDS_MISSING = _("You must recommend a related page OR an external url.")
+    TOO_MANY_FIELDS = _("Please only select a Page OR enter an external url.")
+    MINIMUM_FORMS_MESSAGE = _(
```
To avoid noise in the PR, can you please update these message properties to be `lower_camel_case`, not critical but unless there is a need to change the existing ones we should try to stick with the convention in the file.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -9,15 +9,31 @@
-        # Normalise query string
+        """
+        Save a normalized query string.
+        """
```
there is no need for this docstring change, but if you feel it's important, please align with the current spelling. Normalise, not Normalize.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -91,26 +146,70 @@ class Meta:
+        blank=True,
     )
-    sort_order = models.IntegerField(null=True, blank=True, editable=False)
+    external_link = models.URLField(_("Url"), blank=True)
```
```suggestion
    external_url = models.URLField(_("URL"), blank=True)
```

OR

```suggestion
    external_link_url = models.URLField(_("URL"), blank=True)
```

As noted above, user facing text should use `URL` (it's an abbreviation), and the thing as a whole should be `link`.

lb- commented on wagtail/contrib/search_promotions/models.py:
```diff
@@ -91,26 +146,70 @@ class Meta:
     )
-    sort_order = models.IntegerField(null=True, blank=True, editable=False)
+    external_link = models.URLField(_("Url"), blank=True)
+    external_title = models.CharField(_("Title"), max_length=32, blank=True, null=True)
```
```suggestion
    external_link_title = models.CharField(_("Title"), max_length=32, blank=True, null=True)
```

Only if changing the above one to external_link_url. Not critical though.

lb- commented on wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotion_form.html:
```diff
@@ -10,7 +10,7 @@
     {% endfragment %}
 
-    {% trans "Recommended page" as heading %}
+    {% trans "Search result" as heading %}
```
```suggestion
    {% trans "Recommended search result" as heading %}
```

A reminder that this feature is about Search promotions (aka recommended search results), so changing this to simply 'Search result' is not really correct.

lb- commented on wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/list.html:
```diff
@@ -32,7 +32,11 @@
+                            <a href="{% url 'wagtailadmin_pages:edit' searchpick.page.recommendation.id %}" class="nolink">{{ searchpick.title }}</a>{% if not forloop.last %}, {% endif %}
+                        {% else %}
+                            <a href="{{ searchpick.recommendation.external_link }}" class="nolink">{{ searchpick.title }}</a>{% if not forloop.last %}, {% endif %}
+                        {% endif %}
```
Not sure why but these links are not working for me at all, the href is blank, can you please review.

<img width="1278" alt="Screenshot 2023-08-24 at 8 14 20 am" src="https://github.com/wagtail/wagtail/assets/1396140/2725c49d-9aff-4e3d-a18b-054bcf283992">

I made one with a link to google and the other with a Page and the href is just empty for both.

lb- reviewed (changes requested):
@TopDevPros  thank you, I have done a full review running locally now and have some further feedback.

Please let me know how you go with this, overall this feature is looking good, just needs some refinements.

I would strongly prefer we did not include the most popular query change, it's just adding a bit of complexity to this PR. I understand the desire to do this but if at all possible, can you split that work out to a different PR.

lb- commented:
One more thing, the unit tests will need to all pass, I did not look too deep into the CI failures but from what I can see they are related to this PR and not generic problems with the CI runner.

lb- commented:
Please also include an update to the documentation that covers using the external links. 

https://docs.wagtail.org/en/stable/reference/contrib/searchpromotions.html#usage

lb- commented on docs/reference/contrib/searchpromotions.md:
```diff
@@ -41,12 +41,25 @@ To retrieve a list of promoted search results for a particular search query, you
+                </a>
+            </li>
+        {% endif %}
     {% endfor %}
```
Sorry. Must have missed this. Docs are done. Thanks. 

It may be nicer to put the if/else inside the li element and only output the description once.

TopDevPros commented:
Thanks for the great feedback and for spotting the link bug in wagtailsearchpromotions/list.html. Fixed and a test added to prevent it from happening in the future. We've tried to implement all other requested changes.

Wagtail's "Python coding guidelines" page says to follow PEP 8. PEP 8 is pretty explicit about things like constants. So we restored the original all lower case format for "minimum_forms_message", but use all upper case for new constants.

Thanks again for the thorough and fast feedback! We really appreciate it.

TopDevPros commented:
> One more thing, the unit tests will need to all pass, I did not look too deep into the CI failures but from what I can see they are related to this PR and not generic problems with the CI runner.

We've been unable to duplicate the Assertion error on our systems. We've run all the tests with django 3.2, 4.1, and 4.2; python3.9, 3.10, and 3.11; sqlite 3; and postgresql 12, 13, 14, and 15. No errors appear. 

Note that the test_post_reorder() test fails before the data is reordered. The assertion is just verifying the initial state. This points to a possible database difference. The CI's initial state may be different. 

We've added test code to check in more detail. We've also added logging to capture the database's state before the test begins. It looks like most of the CI test environments pass. We'd like to see the logs from a successful CI run and an unsuccessful one. Hopefully, with those logs and details on the exact environment of each of the associate CI test run, we can better understand where the challenge is.

If this is a CI bug, we'll probably need more access to fix it.

lb- commented on package.json:
```diff
@@ -107,25 +107,45 @@
     "react-redux": "^7.2.2",
     "react-transition-group": "^1.1.3",
+    "readable-stream": "^2.3.8",
+    "readdirp": "^2.2.1",
```
@TopDevPros  did you mean to update this file?

lb- commented on search_promotions.log:
```diff
@@ -0,0 +1,40 @@
+DEBUG:root:   search_pick hello - Root: 1
+DEBUG:root:   editors_picks[1]: hello - Welcome to your new Wagtail site!
+DEBUG:root:   search_pick_2 hello - Welcome to your new Wagtail site!: 2
+WARNING:django.request:Not Found: /admin/searchpicks/
```
Not sure that this log should have been included.

lb- commented:
Thanks @TopDevPros  I have started the CI again, sorry about the comments of the files being included, I read your comment after.

Hopefully the next CI run will let you debug things a bit better.

Please add a comment when ready to review or ready to run the CI again.

TopDevPros commented:
Our analysis suggests that postgresql isn't always refreshing the db between tests. So we've added a fixture to prevent side effects from earlier tests. We're eager to see if this resolves the intermittent failure.  Thank you, @LB!

lb- commented:
I've run the CI again

TopDevPros commented:
> I've run the CI again

 Thanks, lb. Right now, we're rechecking our assumptions.

It would be easier if we could see the error on our systems. Here all the tests pass every time with all the package versions we've tried, which includes the environment that keeps failing in the CI (postgres 12, django 4.1, python 3.10). 

I guess it's good we're stubborn -- excuse me, "determined". :smiley:

TopDevPros commented:
> USE_EMAIL_USER_MODEL=1 ./runtests.py

Thnak you, laymonage, we'll give that a try.

lb- commented:
@TopDevPros  did you mean to close this PR?

TopDevPros commented:
> @TopDevPros did you mean to close this PR?

Oops -- no. 

We are still working on #10725. Apparently we didn't handle the syncing to wagtail's main branch correctly. We'll be more careful in the future.

It appears the real challenge with the test failure isn't directly related to our changes. After a lot of experimentation, we created a new branch from the current wagtail/main repository. We altered the definition of the SearchPromotion's "page" field to accept blank and null. No other changes were made to any other files. We tested that change and got the following error:

ImportError: cannot import name 'CallableChoiceIterator' from 'django.forms.fields' (/opt/hostedtoolcache/Python/3.10.12/x64/lib/python3.10/site-packages/django/forms/fields.py)

Preliminary research suggests this might be a challenge with postgresql and python 3.10. We'll be digging into this in the next few days when time permits.

@lb- we'll be sure to keep you updated.

gasman commented:
@TopDevPros That sounds like you've encountered https://github.com/wagtail/wagtail/pull/10846 - you don't need to worry about that, as it only affects the Django main branch and a fix is already being worked on.

lb- commented:
@TopDevPros ok thanks, sorry about the confusion with the test failures. Hopefully the comments above will help. 

Feel free to raise a new PR against your new branch when ready.

TopDevPros commented:
Thanks @gasman for letting us know about #10846

@ll, we've submitted a new PR  #10852
